In [3]:
import pandas as pd
from collections import Counter

In [63]:
from google.colab import drive

# Mount the Google Drive
drive.mount("/content/drive", force_remount=True)

Mounted at /content/drive


In [5]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [6]:
# Path to the CSV file
file_path = '/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/combined_data_2.csv'

# Read the CSV file into a DataFrame
df = pd.read_csv(file_path)

# Display the first few rows of the DataFrame
df.head(20)

station_id    year  month   day  hour  num_docks_available  \
0          1.0  2022.0   10.0  31.0  23.0             9.000000   
1          1.0  2022.0   11.0   1.0   0.0            11.083333   
2          1.0  2022.0   11.0   1.0   1.0            13.833333   
3          1.0  2022.0   11.0   1.0   2.0            12.666667   
4          1.0  2022.0   11.0   1.0   3.0            10.083333   
5          1.0  2022.0   11.0   1.0   4.0             8.916667   
6          1.0  2022.0   11.0   1.0   5.0            10.833333   
7          1.0  2022.0   11.0   1.0   6.0            10.916667   
8          1.0  2022.0   11.0   1.0   7.0            10.666667   
9          1.0  2022.0   11.0   1.0   8.0            12.833333   
10         1.0  2022.0   11.0   1.0   9.0            13.583333   
11         1.0  2022.0   11.0   1.0  10.0            15.500000   
12         1.0  2022.0   11.0   1.0  11.0            15.666667   
13         1.0  2022.0   11.0   1.0  12.0            18.083333   
14         1.0  2022.0   11.0   1.0  13.0            21.000000   
15         1.0  2022.0   11.0   1.0  14.0            19.333333   
16         1.0  2022.0   11.0   1.0  15.0            22.166667   
17         1.0  2022.0   11.0   1.0  16.0            24.166667   
18         1.0  2022.0   11.0   1.0  17.0            23.833333   
19         1.0  2022.0   11.0   1.0  18.0            25.333333   

    num_bikes_available  
0             35.000000  
1             33.750000  
2             31.166667  
3             32.250000  
4             34.833333  
5             35.916667  
6             34.166667  
7             34.083333  
8             34.333333  
9             32.166667  
10            31.333333  
11            29.416667  
12            28.583333  
13            26.583333  
14            23.000000  
15            24.500000  
16            21.750000  
17            19.833333  
18            19.666667  
19            18.666667

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16549540 entries, 0 to 16549539
Data columns (total 7 columns):
 #   Column               Dtype  
---  ------               -----  
 0   station_id           float64
 1   year                 float64
 2   month                float64
 3   day                  float64
 4   hour                 float64
 5   num_docks_available  float64
 6   num_bikes_available  float64
dtypes: float64(7)
memory usage: 883.8 MB


Cuando hice la media, se cambió el formato a float y lo voy a poner a entero.
La media la calcule en el archivo Bike_kaggel_1_Availability_Prediction.ipynb

In [8]:
df['station_id'] = df['station_id'].astype(int)
df['year'] = df['year'].astype(int)
df['month'] = df['month'].astype(int)
df['day'] = df['day'].astype(int)
df['hour'] = df['hour'].astype(int)

In [9]:
df

station_id  year  month  day  hour  num_docks_available  \
0                  1  2022     10   31    23             9.000000   
1                  1  2022     11    1     0            11.083333   
2                  1  2022     11    1     1            13.833333   
3                  1  2022     11    1     2            12.666667   
4                  1  2022     11    1     3            10.083333   
...              ...   ...    ...  ...   ...                  ...   
16549535         519  2021      9   30    20            20.666667   
16549536         519  2021      9   30    21            19.000000   
16549537         519  2021      9   30    22            17.833333   
16549538         519  2021      9   30    23            17.000000   
16549539         519  2021     10    1     0            17.000000   

          num_bikes_available  
0                   35.000000  
1                   33.750000  
2                   31.166667  
3                   32.250000  
4                   34.833333  
...                       ...  
16549535             2.333333  
16549536             4.000000  
16549537             5.166667  
16549538             6.000000  
16549539             6.000000  

[16549540 rows x 7 columns]

Calcule una lista con los ID de las estaciones de la base de datos completa.

In [10]:
counter_ids = Counter(df['station_id'])
stations_database = list(counter_ids.keys())

Calcule una lista con los ID de las estaciones del archivo metadata_sample_submission_2024.

In [11]:
filepath_metadata = '/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/metadata_sample_submission_2024.csv'
df_metadata = pd.read_csv(filepath_metadata, sep=',')

In [12]:
counter_ids_sub = Counter(df_metadata['station_id'])
stations_submission = list(counter_ids_sub.keys())

In [13]:
len(stations_database)

523

In [14]:
len(stations_submission)

399

Tengo 399 estaciones que son las que necesito predecir, pero un total de 529 en la base de datos completa, por lo que voy a eliminar muchas estaciones que no voy a utilizar.

In [15]:
for station in stations_submission:
  if station not in stations_database:
    print(f'station {station} is not in you database')

In [16]:
stations_to_remove = []
for station in stations_database:
  if station not in stations_submission:
    print(f'You can remove station {station} from your database')
    stations_to_remove.append(station)

You can remove station 10 from your database
You can remove station 32 from your database
You can remove station 37 from your database
You can remove station 43 from your database
You can remove station 48 from your database
You can remove station 52 from your database
You can remove station 66 from your database
You can remove station 77 from your database
You can remove station 91 from your database
You can remove station 188 from your database
You can remove station 211 from your database
You can remove station 234 from your database
You can remove station 245 from your database
You can remove station 287 from your database
You can remove station 311 from your database
You can remove station 330 from your database
You can remove station 355 from your database
You can remove station 366 from your database
You can remove station 379 from your database
You can remove station 383 from your database
You can remove station 399 from your database
You can remove station 403 from your databa

Elimino las estaciones que NO me van a hacer falta.

In [17]:
df_filtered = df[~df['station_id'].isin(stations_to_remove)]
df_filtered

station_id  year  month  day  hour  num_docks_available  \
0                  1  2022     10   31    23             9.000000   
1                  1  2022     11    1     0            11.083333   
2                  1  2022     11    1     1            13.833333   
3                  1  2022     11    1     2            12.666667   
4                  1  2022     11    1     3            10.083333   
...              ...   ...    ...  ...   ...                  ...   
16533034         496  2021      9   30    20            10.666667   
16533035         496  2021      9   30    21             6.833333   
16533036         496  2021      9   30    22             3.166667   
16533037         496  2021      9   30    23             6.666667   
16533038         496  2021     10    1     0             4.333333   

          num_bikes_available  
0                   35.000000  
1                   33.750000  
2                   31.166667  
3                   32.250000  
4                   34.833333  
...                       ...  
16533034             1.666667  
16533035             5.500000  
16533036             8.833333  
16533037             5.750000  
16533038             7.000000  

[13208394 rows x 7 columns]

Voy a obtener la capacidad de cada estación a partir del archivo de informació_Estacions_Bicing.csv

In [18]:
filepath_info = '/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/Informacio_Estacions_Bicing.csv'

In [19]:
df_info = pd.read_csv(filepath_info, sep=',')
df_info

station_id                                 name physical_configuration  \
0             1        GRAN VIA CORTS CATALANES, 760    ELECTRICBIKESTATION   
1             2                C/ ROGER DE FLOR, 126    ELECTRICBIKESTATION   
2             3                        C/ NÀPOLS, 82    ELECTRICBIKESTATION   
3             4                         C/ RIBES, 13    ELECTRICBIKESTATION   
4             5  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)    ELECTRICBIKESTATION   
..          ...                                  ...                    ...   
501         515                    C/ SANT ADRIÀ, 43    ELECTRICBIKESTATION   
502         516                    C/ SANT ADRIÀ, 88    ELECTRICBIKESTATION   
503         517             AV. RASOS DE PEGUERA, 10    ELECTRICBIKESTATION   
504         518                    C/ LLOBREGÓS, 115    ELECTRICBIKESTATION   
505         519                       C/ PEDRELL, 52    ELECTRICBIKESTATION   

           lat       lon  altitude                              address  \
0    41.397978  2.180107      16.0        GRAN VIA CORTS CATALANES, 760   
1    41.395488  2.177198      17.0                C/ ROGER DE FLOR, 126   
2    41.394156  2.181331      11.0                        C/ NÀPOLS, 82   
3    41.393317  2.181248       8.0                         C/ RIBES, 13   
4    41.391103  2.180176       7.0  PG. LLUIS COMPANYS, 11 (ARC TRIOMF)   
..         ...       ...       ...                                  ...   
501  41.435207  2.194800      19.0                    C/ SANT ADRIÀ, 43   
502  41.435460  2.200157      15.0                    C/ SANT ADRIÀ, 88   
503  41.462095  2.178959      44.0             AV. RASOS DE PEGUERA, 10   
504  41.424689  2.157049     112.0                    C/ LLOBREGÓS, 115   
505  41.424655  2.166289     110.0                       C/ PEDRELL, 52   

     post_code  capacity  is_charging_station  nearby_distance  \
0         8013        45                 True           1000.0   
1         8013        29                 True           1000.0   
2         8013        27                 True           1000.0   
3         8013        21                 True           1000.0   
4         8018        39                 True           1000.0   
..         ...       ...                  ...              ...   
501       8030        24                 True           1000.0   
502       8030        21                 True           1000.0   
503       8033        20                 True           1000.0   
504       8032        27                 True           1000.0   
505       8032        24                 True           1000.0   

     _ride_code_support  rental_uris cross_street  
0                  True          NaN          NaN  
1                  True          NaN          NaN  
2                  True          NaN          NaN  
3                  True          NaN          NaN  
4                  True          NaN          NaN  
..                  ...          ...          ...  
501                True          NaN          NaN  
502                True          NaN          NaN  
503                True          NaN          NaN  
504                True          NaN          NaN  
505                True          NaN          NaN  

[506 rows x 14 columns]

In [20]:
df_info_simplified = df_info[['station_id', 'capacity']]
df_info_simplified

station_id  capacity
0             1        45
1             2        29
2             3        27
3             4        21
4             5        39
..          ...       ...
501         515        24
502         516        21
503         517        20
504         518        27
505         519        24

[506 rows x 2 columns]

In [21]:
len(df_info_simplified)

506

Tengo la información de la capacidad de 506 estaciones en total, pero solo necesito 399.

In [22]:
counter_ids = Counter(df_filtered['station_id'])
stations_database = list(counter_ids.keys())

In [23]:
stations_info = list(df_info_simplified['station_id'])

In [24]:
missing_stations = []
for station in stations_database:
  if station not in stations_info:
    print(f"Station {station} is missing")
    missing_stations.append(station)

No falta información de la capacidad de ninguna estación de las que tengo que hacer predicciones. Por lo que voy a añadir una nueva columna en la base de datos con la información de la capacidad.

Primero elimino la columna que NO me va a hacer falta.
Numero de muelles + capacidad

In [25]:
df_filtered.drop('num_bikes_available', axis=1, inplace=True)
df_filtered.head()

<ipython-input-25-020564ee2d67>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered.drop('num_bikes_available', axis=1, inplace=True)


station_id  year  month  day  hour  num_docks_available
0           1  2022     10   31    23             9.000000
1           1  2022     11    1     0            11.083333
2           1  2022     11    1     1            13.833333
3           1  2022     11    1     2            12.666667
4           1  2022     11    1     3            10.083333

Voy a dibujar algo para ubicarme como es que debo organizar las filas, para el analisis, aqui estoy un poco confundido. Cómo convertir 5 filas en 1 (X, y)

    x =     1  2020      3    1     3      16.00  13.00    6.33   4.58      y= 2.1666                                                 
  

Si falta una hora la voy a considerar que tome la siguiente que exista.
Como explicaba Pere en el correo de gmail.

Voy a ordenar los datos por estación, año, mes, día y hora.

In [26]:
df_ordenado = df_filtered.sort_values(by=['station_id', 'year', 'month', 'day', 'hour'], ascending=[True, True, True, True, True])
df_ordenado

station_id  year  month  day  hour  num_docks_available
7207390           1  2019     12   31    23            23.000000
7207391           1  2020      1    1     0            22.166667
7207392           1  2020      1    1     1            20.666667
7207393           1  2020      1    1     2            17.750000
7207394           1  2020      1    1     3            15.583333
...             ...   ...    ...  ...   ...                  ...
3142603         496  2023     12   31    20            11.583333
3142604         496  2023     12   31    21            13.250000
3142605         496  2023     12   31    22            13.166667
3142606         496  2023     12   31    23            12.916667
3142607         496  2024      1    1     0            11.666667

[13208394 rows x 6 columns]

Adiciono, la columna capacidad para calcular el porcentaje de docks disponibles.

Asi se accedo a la capacidad de una estación concreta con el archivo info:

In [27]:
id = 2
df_info_simplified[df_info_simplified['station_id']==id]['capacity'].iloc[0]

29

In [28]:
# Integracion
def add_capacity(row):
    id = row['station_id']
    return df_info_simplified[df_info_simplified['station_id']==id]['capacity'].iloc[0]

# Usar apply para aplicar la función a cada fila y crear la nueva columna
df_ordenado['capacity'] = df_ordenado.apply(add_capacity, axis=1)
df_ordenado

station_id  year  month  day  hour  num_docks_available  capacity
7207390           1  2019     12   31    23            23.000000        45
7207391           1  2020      1    1     0            22.166667        45
7207392           1  2020      1    1     1            20.666667        45
7207393           1  2020      1    1     2            17.750000        45
7207394           1  2020      1    1     3            15.583333        45
...             ...   ...    ...  ...   ...                  ...       ...
3142603         496  2023     12   31    20            11.583333        18
3142604         496  2023     12   31    21            13.250000        18
3142605         496  2023     12   31    22            13.166667        18
3142606         496  2023     12   31    23            12.916667        18
3142607         496  2024      1    1     0            11.666667        18

[13208394 rows x 7 columns]

In [29]:
df_ordenado.to_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/preprocessed_data.csv', index=False)

In [30]:
df_ordenado = pd.read_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/preprocessed_data.csv')

In [31]:
df_ordenado

station_id  year  month  day  hour  num_docks_available  capacity
0                  1  2019     12   31    23            23.000000        45
1                  1  2020      1    1     0            22.166667        45
2                  1  2020      1    1     1            20.666667        45
3                  1  2020      1    1     2            17.750000        45
4                  1  2020      1    1     3            15.583333        45
...              ...   ...    ...  ...   ...                  ...       ...
13208389         496  2023     12   31    20            11.583333        18
13208390         496  2023     12   31    21            13.250000        18
13208391         496  2023     12   31    22            13.166667        18
13208392         496  2023     12   31    23            12.916667        18
13208393         496  2024      1    1     0            11.666667        18

[13208394 rows x 7 columns]

Calcule la **Proporcion** de muelles disponibles = el numero de muelles disponibles / la capacidad.

In [32]:
 # Esta es mi variable a predecir
 # No confundir Numero de muelles disponibles con la Proporcion de muelles disponibles, son dos cosas diferentes.
 # 20 capacidad de muelles en cada estacion y 10 es la disponibilidad de muelles en cada estacion = 0.5 proporcion de muelles
df_ordenado['percentage_docks_available'] = df_ordenado['num_docks_available'] / df_ordenado['capacity']

In [33]:
df_ordenado

station_id  year  month  day  hour  num_docks_available  capacity  \
0                  1  2019     12   31    23            23.000000        45   
1                  1  2020      1    1     0            22.166667        45   
2                  1  2020      1    1     1            20.666667        45   
3                  1  2020      1    1     2            17.750000        45   
4                  1  2020      1    1     3            15.583333        45   
...              ...   ...    ...  ...   ...                  ...       ...   
13208389         496  2023     12   31    20            11.583333        18   
13208390         496  2023     12   31    21            13.250000        18   
13208391         496  2023     12   31    22            13.166667        18   
13208392         496  2023     12   31    23            12.916667        18   
13208393         496  2024      1    1     0            11.666667        18   

          percentage_docks_available  
0                           0.511111  
1                           0.492593  
2                           0.459259  
3                           0.394444  
4                           0.346296  
...                              ...  
13208389                    0.643519  
13208390                    0.736111  
13208391                    0.731481  
13208392                    0.717593  
13208393                    0.648148  

[13208394 rows x 8 columns]

Aqui voy a eliminar las columnas que ya NO me van a hacer falta, por que ya tengo el Percentage_docks_available.

In [34]:
df_ordenado.drop('capacity', axis=1, inplace=True)
df_ordenado.drop('num_docks_available', axis=1, inplace=True)

df_ordenado.head()

station_id  year  month  day  hour  percentage_docks_available
0           1  2019     12   31    23                    0.511111
1           1  2020      1    1     0                    0.492593
2           1  2020      1    1     1                    0.459259
3           1  2020      1    1     2                    0.394444
4           1  2020      1    1     3                    0.346296

In [35]:
df_n_station = df_ordenado[df_ordenado['station_id'] == 1]
df_n_station.head(50)

station_id  year  month  day  hour  percentage_docks_available
0            1  2019     12   31    23                    0.511111
1            1  2020      1    1     0                    0.492593
2            1  2020      1    1     1                    0.459259
3            1  2020      1    1     2                    0.394444
4            1  2020      1    1     3                    0.346296
5            1  2020      1    1     4                    0.283333
6            1  2020      1    1     5                    0.364815
7            1  2020      1    1     6                    0.298148
8            1  2020      1    1     7                    0.233333
9            1  2020      1    1     8                    0.235185
10           1  2020      1    1     9                    0.248148
11           1  2020      1    1    10                    0.262963
12           1  2020      1    1    11                    0.342593
13           1  2020      1    1    12                    0.337037
14           1  2020      1    1    13                    0.385185
15           1  2020      1    1    14                    0.381481
16           1  2020      1    1    15                    0.335185
17           1  2020      1    1    16                    0.244444
18           1  2020      1    1    17                    0.211111
19           1  2020      1    1    18                    0.177778
20           1  2020      1    1    19                    0.079630
21           1  2020      1    1    20                    0.209259
22           1  2020      1    1    21                    0.418519
23           1  2020      1    1    22                    0.492593
24           1  2020      1    1    23                    0.450000
25           1  2020      1    2     0                    0.420370
26           1  2020      1    2     1                    0.387037
27           1  2020      1    2     2                    0.422222
28           1  2020      1    2     3                    0.422222
29           1  2020      1    2     4                    0.422222
30           1  2020      1    2     5                    0.422222
31           1  2020      1    2     6                    0.431481
32           1  2020      1    2     7                    0.546296
33           1  2020      1    2     8                    0.611111
34           1  2020      1    2     9                    0.640741
35           1  2020      1    2    10                    0.705556
36           1  2020      1    2    11                    0.751852
37           1  2020      1    2    12                    0.798148
38           1  2020      1    2    13                    0.733333
39           1  2020      1    2    14                    0.731481
40           1  2020      1    2    15                    0.720370
41           1  2020      1    2    16                    0.742593
42           1  2020      1    2    17                    0.685185
43           1  2020      1    2    18                    0.650000
44           1  2020      1    2    19                    0.451852
45           1  2020      1    2    20                    0.346296
46           1  2020      1    2    21                    0.275926
47           1  2020      1    2    22                    0.151852
48           1  2020      1    2    23                    0.022222
49           1  2020      1    3     0                    0.055556

Lo siguiente es calcular los contextos de mis datos (CXT-4, CXT-3, CXT-2, CXT-1) estos son los contextos.
Con los anteriores contextos tengo que coger agrupaciones de 5 filas seguidas, segun instrucciones  de PERE. "Corresponderian a 5 horas seguidas, de informacion sobre/en una estacion"
***Las filas las voy a tomar de 5 en 5 por station_id de c/d estacion, por eso hay que segmentar las estaciones primero.

Calcule los contextos para todas las estaciones y sus datos. Para ello primero segmente por cada estación y cree una lista con todas las estaciones disponibles

In [36]:
# Lista de estaciones
counter_ids_filtered = Counter(df_ordenado['station_id'])
stations_filtered = list(counter_ids_filtered.keys())
stations_filtered

[1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 11,
 12,
 13,
 14,
 15,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 33,
 34,
 35,
 36,
 39,
 40,
 41,
 42,
 44,
 45,
 46,
 47,
 49,
 50,
 51,
 53,
 54,
 55,
 56,
 57,
 58,
 60,
 61,
 62,
 63,
 64,
 65,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 92,
 94,
 95,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 170,
 171,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 182,
 183,
 184,
 185,
 186,
 187,
 189,
 190,
 192,
 193,
 194,
 195,
 196,
 197,
 198,
 199,
 200,
 201,
 202,
 2

In [37]:
# Comprobe que no me hace falta ninguna estacion
len(stations_filtered)

399

En el primer Bucle recorri todas las estaciones.    
En el segundo Bucle, recorri todas las filas de 5 en 5 de una estación concreta para calcular todos los contextos.

In [38]:
result_list = []
for station_id in stations_filtered:  # Bucle, en cada iteracion se procesa 1 station_id
  print(f'Analyzing station {station_id}')
  df_n_station = df_ordenado[df_ordenado['station_id'] == station_id] # Cogi todos los datos de station_id
  for i in range(0, len(df_n_station), 5):  # Bucle, recorri todos los datos de esta estacion de 5 en 5
    chunk =  df_n_station.iloc[i:i+5].copy()  # Cogi las 5 filas siguientes
    if len(chunk) < 5:  # Si el numero de filas no es múltiplo de 5, ignoramos las ultimas restantes que queden solas. Es decir, no las añadimos a nuestra base de datos.
      continue
    new_lin = chunk.iloc[4].copy()
    new_lin['ctx-4'] = chunk.iloc[0]['percentage_docks_available']
    new_lin['ctx-3'] = chunk.iloc[1]['percentage_docks_available']
    new_lin['ctx-2'] = chunk.iloc[2]['percentage_docks_available']
    new_lin['ctx-1'] = chunk.iloc[3]['percentage_docks_available']
    result_list.append(new_lin)

# Genere un a lista y los converti en Dataframe
result_df = pd.DataFrame(result_list)
result_df.head()

Analyzing station 1
Analyzing station 2
Analyzing station 3
Analyzing station 4
Analyzing station 5
Analyzing station 6
Analyzing station 7
Analyzing station 8
Analyzing station 9
Analyzing station 11
Analyzing station 12
Analyzing station 13
Analyzing station 14
Analyzing station 15
Analyzing station 17
Analyzing station 18
Analyzing station 19
Analyzing station 20
Analyzing station 21
Analyzing station 22
Analyzing station 23
Analyzing station 24
Analyzing station 25
Analyzing station 26
Analyzing station 27
Analyzing station 28
Analyzing station 29
Analyzing station 30
Analyzing station 31
Analyzing station 33
Analyzing station 34
Analyzing station 35
Analyzing station 36
Analyzing station 39
Analyzing station 40
Analyzing station 41
Analyzing station 42
Analyzing station 44
Analyzing station 45
Analyzing station 46
Analyzing station 47
Analyzing station 49
Analyzing station 50
Analyzing station 51
Analyzing station 53
Analyzing station 54
Analyzing station 55
Analyzing station 56
A

station_id    year  month  day  hour  percentage_docks_available  \
4          1.0  2020.0    1.0  1.0   3.0                    0.346296   
9          1.0  2020.0    1.0  1.0   8.0                    0.235185   
14         1.0  2020.0    1.0  1.0  13.0                    0.385185   
19         1.0  2020.0    1.0  1.0  18.0                    0.177778   
24         1.0  2020.0    1.0  1.0  23.0                    0.450000   

       ctx-4     ctx-3     ctx-2     ctx-1  
4   0.511111  0.492593  0.459259  0.394444  
9   0.283333  0.364815  0.298148  0.233333  
14  0.248148  0.262963  0.342593  0.337037  
19  0.381481  0.335185  0.244444  0.211111  
24  0.079630  0.209259  0.418519  0.492593

In [39]:
result_df.to_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/preprocessed_data_ctxs.csv', index=False)

In [40]:
result_df = pd.read_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/preprocessed_data_ctxs.csv')
result_df

station_id    year  month   day  hour  percentage_docks_available  \
0               1.0  2020.0    1.0   1.0   3.0                    0.346296   
1               1.0  2020.0    1.0   1.0   8.0                    0.235185   
2               1.0  2020.0    1.0   1.0  13.0                    0.385185   
3               1.0  2020.0    1.0   1.0  18.0                    0.177778   
4               1.0  2020.0    1.0   1.0  23.0                    0.450000   
...             ...     ...    ...   ...   ...                         ...   
2641510       496.0  2023.0   12.0  31.0   3.0                    0.504630   
2641511       496.0  2023.0   12.0  31.0   8.0                    0.388889   
2641512       496.0  2023.0   12.0  31.0  13.0                    0.861111   
2641513       496.0  2023.0   12.0  31.0  18.0                    0.666667   
2641514       496.0  2023.0   12.0  31.0  23.0                    0.717593   

            ctx-4     ctx-3     ctx-2     ctx-1  
0        0.511111  0.492593  0.459259  0.394444  
1        0.283333  0.364815  0.298148  0.233333  
2        0.248148  0.262963  0.342593  0.337037  
3        0.381481  0.335185  0.244444  0.211111  
4        0.079630  0.209259  0.418519  0.492593  
...           ...       ...       ...       ...  
2641510  0.564815  0.518519  0.606481  0.523148  
2641511  0.555556  0.537037  0.490741  0.435185  
2641512  0.490741  0.412037  0.551282  0.646465  
2641513  0.879630  0.819444  0.768519  0.699074  
2641514  0.722222  0.643519  0.736111  0.731481  

[2641515 rows x 10 columns]

Ya tengo la base de datos procesada, solo me hace falta dividir en X e Y, siendo **(X)** el 'metadata_sample_submission_2024.csv' y **(Y)** los valores a predecir por el modelo.

Voy a ordenar por fecha para hacer los splits de Train-Validation-Test en el orden natural de los datos y evitar data linking que Pere menciono en el video.

In [41]:
result_df_ordenado = result_df.sort_values(by=['year', 'month', 'day', 'hour'], ascending=[True, True, True, True])

In [42]:
X = result_df_ordenado.drop(columns=['percentage_docks_available']) # Esto refleja la informacion de todos los Contextos con la Fecha.
Y = result_df_ordenado['percentage_docks_available'] # Es la informacion de la Proporcion de muelles disponibles

In [43]:
X.head()

station_id    year  month  day  hour     ctx-4     ctx-3     ctx-2  \
0             1.0  2020.0    1.0  1.0   3.0  0.511111  0.492593  0.459259   
6624          2.0  2020.0    1.0  1.0   3.0  0.551724  0.571839  0.689655   
13141         3.0  2020.0    1.0  1.0   3.0  0.333333  0.379630  0.388889   
19786         4.0  2020.0    1.0  1.0   3.0  0.619048  0.650794  0.773810   
26425         5.0  2020.0    1.0  1.0   3.0  0.615385  0.630342  0.638889   

          ctx-1  
0      0.394444  
6624   0.859195  
13141  0.361111  
19786  0.718254  
26425  0.688034

In [44]:
Y.head()

0        0.346296
6624     0.816092
13141    0.395062
19786    0.753968
26425    0.692308
Name: percentage_docks_available, dtype: float64

Divido el Train-Val-Test, los datos ordenados por fecha

In [45]:
import  numpy as np

In [46]:
X_array = np.array(X)
Y_array = np.array(Y)

In [47]:
# Get the total number of samples
n_samples = len(X_array)

# Define the split sizes
train_size = int(0.8 * n_samples)  # 80% for training
val_size = n_samples - train_size  # Remaining for validation

In [48]:
# Split the data without shuffling
X_train = X_array[:train_size]
Y_train = Y_array[:train_size]

X_val = X_array[train_size:]
Y_val = Y_array[train_size:]

print("Training set size:", len(X_train))
print("Validation set size:", len(X_val))


Training set size: 2113212
Validation set size: 528303


Obtuve el X_test del archivo 'metadata_sample_submission_2024.csv'

In [49]:
filepath_metadata = '/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/metadata_sample_submission_2024.csv'
df_metadata = pd.read_csv(filepath_metadata, sep=',')
df_metadata

index  station_id  month  day  hour     ctx-4     ctx-3     ctx-2  \
0            0           1      1    1     5  0.781481  0.677778  0.696296   
1            1           1      1    1    10  0.737374  0.711111  0.711111   
2            2           1      1    1    15  0.827778  0.896296  0.901852   
3            3           1      1    1    20  0.825926  0.874074  0.927778   
4            4           2      1    1     3  0.592593  0.341954  0.275862   
...        ...         ...    ...  ...   ...       ...       ...       ...   
171897  171897         496      3   31     1  0.828704  0.787037  0.777778   
171898  171898         496      3   31     6  0.930556  0.944444  0.935185   
171899  171899         496      3   31    11  0.912037  0.884259  0.518519   
171900  171900         496      3   31    16  0.245370  0.319444  0.277778   
171901  171901         496      3   31    21  0.513889  0.435185  0.490741   

           ctx-1  
0       0.750000  
1       0.731624  
2       0.883333  
3       0.918519  
4       0.540230  
...          ...  
171897  0.853535  
171898  0.856481  
171899  0.157407  
171900  0.305556  
171901  0.504630  

[171902 rows x 9 columns]

Añadimos la columna del año 2024

In [50]:
df_metadata['year'] = 2024
X_test = df_metadata[['station_id', 'year', 'month', 'day', 'hour', 'ctx-4', 'ctx-3', 'ctx-2', 'ctx-1']]
X_test

station_id  year  month  day  hour     ctx-4     ctx-3     ctx-2  \
0                1  2024      1    1     5  0.781481  0.677778  0.696296   
1                1  2024      1    1    10  0.737374  0.711111  0.711111   
2                1  2024      1    1    15  0.827778  0.896296  0.901852   
3                1  2024      1    1    20  0.825926  0.874074  0.927778   
4                2  2024      1    1     3  0.592593  0.341954  0.275862   
...            ...   ...    ...  ...   ...       ...       ...       ...   
171897         496  2024      3   31     1  0.828704  0.787037  0.777778   
171898         496  2024      3   31     6  0.930556  0.944444  0.935185   
171899         496  2024      3   31    11  0.912037  0.884259  0.518519   
171900         496  2024      3   31    16  0.245370  0.319444  0.277778   
171901         496  2024      3   31    21  0.513889  0.435185  0.490741   

           ctx-1  
0       0.750000  
1       0.731624  
2       0.883333  
3       0.918519  
4       0.540230  
...          ...  
171897  0.853535  
171898  0.856481  
171899  0.157407  
171900  0.305556  
171901  0.504630  

[171902 rows x 9 columns]

In [51]:
X_test = np.array(X_test)


Ahora voy a probar el modelo de regresión propuesto por Pere en su correo.

In [52]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [53]:
# Initialize the model
model = LinearRegression()

# Train the model on the training set
model.fit(X_train, Y_train)

# Predict on the validation set
Y_val_pred = model.predict(X_val)

# Evaluate the model on the validation set
mse_val = mean_squared_error(Y_val, Y_val_pred)
r2_val = r2_score(Y_val, Y_val_pred)

print("Validation MSE:", mse_val)
print("Validation R^2:", r2_val)

Validation MSE: 0.011463574818539077
Validation R^2: 0.84144134911256


Ya que le modelo está entrenado, ahora voy a calcular las predicciones para X_test de kaggle.

In [54]:
# Predict on the test set
Y_test_pred = model.predict(X_test)

In [55]:
Y_test_pred

array([0.73575846, 0.71433879, 0.84291827, ..., 0.14583734, 0.3497019 ,
       0.51576832])

Voy a guardar los resultados en el formato indicado en la competencia de Kaggel.

In [56]:
df_submission = pd.DataFrame({
    'index': df_metadata['index'],
    'percentage_docks_available': Y_test_pred
})

In [57]:
df_submission

index  percentage_docks_available
0            0                    0.735758
1            1                    0.714339
2            2                    0.842918
3            3                    0.873107
4            4                    0.595710
...        ...                         ...
171897  171897                    0.847214
171898  171898                    0.816645
171899  171899                    0.145837
171900  171900                    0.349702
171901  171901                    0.515768

[171902 rows x 2 columns]

In [58]:
df_submission.to_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/linear_regression_submission_2024.csv', index=False)

Voy a probar un segundo modelo.

In [59]:
from sklearn.ensemble import RandomForestRegressor


In [60]:
# Initialize the model
model_rf = RandomForestRegressor(n_estimators=100, random_state=42)

model_rf.fit(X_train, Y_train)

Y_val_pred_rf = model_rf.predict(X_val)

# Evaluate the model on the validation set
mse_val_rf = mean_squared_error(Y_val, Y_val_pred_rf)
r2_val_rf = r2_score(Y_val, Y_val_pred_rf)

print("Validation MSE:", mse_val_rf)
print("Validation R^2:", r2_val_rf)

# Predict on the test set
Y_test_pred_rf = model.predict(X_test)

Validation MSE: 0.010902581637765741
Validation R^2: 0.849200736852292


In [61]:
df_submission_rf = pd.DataFrame({
    'index': df_metadata['index'],
    'percentage_docks_available': Y_test_pred
})

In [62]:
df_submission_rf.to_csv('/content/drive/MyDrive/University of Barcelona/Capstone_Project_2024/random_forest_submission_2024.csv', index=False)